diffkemp/semdiff/caching.py

```diff
@@ -0,0 +1,636 @@
+            for nonfun_diff in vertex.nonfun_diffs:
+                nonfun_pair = []
+                for side in ComparisonGraph.Side:
+                    syndiff_bodies = (syndiff_bodies_left
+                                      if side == ComparisonGraph.Side.LEFT
+                                      else syndiff_bodies_right)
+                    backtracking_map = (backtracking_map_left
+                                        if side == ComparisonGraph.Side.LEFT
+                                        else backtracking_map_right)
+                    # Convert the YAML callstack format to string.
+                    calls = self._yaml_callstack_to_string(
+                        nonfun_diff.callstack[side])
+                    # Append the parent function's callstack.
+                    # (unless it is the base function)
+                    fun = fun_first if side == ComparisonGraph.Side.LEFT \
+                        else fun_second
+                    if nonfun_diff.parent_fun != fun:
+                        parent_calls = self._edge_callstack_to_string(
+                            _get_callstack(backtracking_map, self[fun],
+                                           vertex))
+                        calls = parent_calls + "\n" + calls
+
+                    if isinstance(nonfun_diff, ComparisonGraph.SyntaxDiff):
+                        nonfun_pair.append(Result.Entity(
+                            nonfun_diff.name,
+                            None,
+                            None,
+                            calls,
+                            "syntactic",
+                            False
+                        ))
+                        syndiff_bodies[nonfun_diff.name] = \
+                            nonfun_diff.body[side]
+                    elif isinstance(nonfun_diff, ComparisonGraph.TypeDiff):
+                        nonfun_pair.append(Result.Entity(
+                            nonfun_diff.name,
+                            nonfun_diff.file[side],
+                            nonfun_diff.line[side],
+                            calls,
+                            "type",
+                            False
+                        ))
+                # Non-function differences are always of the non-equal type
+                nonfun_pair.append(Result.Kind.NOT_EQUAL)
+                objects_to_compare.append(tuple(nonfun_pair))
+        return objects_to_compare, syndiff_bodies_left, syndiff_bodies_right
+
+
+def _get_callstack(backtracking_map, start_vertex, end_vertex):
+    """Generates an edge callstack based on the backtracking map."""
+    if start_vertex == end_vertex:
+        return []
+    edges = [backtracking_map[end_vertex]]
+    while edges[-1].parent_vertex is not start_vertex:
+        edges.append(backtracking_map[edges[-1].parent_vertex])
+    return list(reversed(edges))
+
+
+class SimpLLCache:
+    """
+    A counterpart to ComparisonGraph responsible for providing the required
```

Comment:
This is not really a "counterpart", it's more like a _class responsible for providing the results stored in the comparison graph to SimpLL in the form of a cache so that SimpLL does not have to re-compare functions that have been compared before_
or _a class responsible for generating a cache from the comparison result that SimpLL can use to prevent repeated comparison of functions that have been compared before_